debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/RecordsSnapshotProducer.java

```diff
@@ -176,48 +177,75 @@ private void takeSnapshot(Consumer<ChangeEvent> consumer) {
             logger.info("Step 3: reading and exporting the contents of each table");
             AtomicInteger rowsCounter = new AtomicInteger(0);
             final Map<TableId, String> selectOverrides = getSnapshotSelectOverridesByTable();
-            schema.tables().forEach(tableId -> {
+
+            // whether this thread has been interrupted (i.e. the connector has been stopped before the snapshot was completed)
+            boolean interrupted = false;
+
+            for(TableId tableId : schema.tables()) {
+                if (interrupted) {
+                    // We were not able to finish all rows in all tables ...
+                    logger.warn("Aborting the snapshot due to thread interruption before all tables were processed");
+                    break;
+                }
                 if (schema.isFilteredOut(tableId)) {
                     logger.info("\t table '{}' is filtered out, ignoring", tableId);
-                    return;
+                    continue;
                 }
+
                 long exportStart = clock().currentTimeInMillis();
                 logger.info("\t exporting data from table '{}'", tableId);
                 try {
                     // DBZ-298 Quoting name in case it has been quoted originally; it doesn't do harm if it hasn't been quoted
                     final String selectStatement = selectOverrides.getOrDefault(tableId, "SELECT * FROM " + tableId.toDoubleQuotedString());
                     logger.info("For table '{}' using select statement: '{}'", tableId, selectStatement);
 
-                    connection.query(selectStatement,
-                                     this::readTableStatement,
-                                     rs -> readTable(tableId, rs, consumer, rowsCounter));
+                    connection.queryWithBlockingConsumer(selectStatement,
+                            this::readTableStatement,
+                            rs -> readTable(tableId, rs, consumer, rowsCounter));
                     logger.info("\t finished exporting '{}' records for '{}'; total duration '{}'", rowsCounter.get(),
-                                tableId, Strings.duration(clock().currentTimeInMillis() - exportStart));
+                            tableId, Strings.duration(clock().currentTimeInMillis() - exportStart));
                     rowsCounter.set(0);
                 } catch (SQLException e) {
                     throw new ConnectException(e);
                 }
-            });
+                catch (InterruptedException e) {
+                    Thread.interrupted();
+                    interrupted = true;
+                }
+            }
 
             // finally commit the transaction to release all the locks...
             logger.info("Step 4: committing transaction '{}'", txId);
             jdbcConnection.commit();
 
```

Comment:
We can terminate the loop here
```java
if (Thread.isInterrupted()) {
    break;
}
````

Reply:
I've pulled out that catch now. Let me know what you think.